### PR TITLE
feat: add tray and menu javascript APIs and events, closes #6617

### LINKED ISSUE #6617: [feat] JS window menu API
### Describe the problem

Window menu can't be set from JS.

### Describe the solution you'd like

A JS window menu API would allow setting window menu from JS.

### Alternatives considered

None

### Additional context

None

---
Thanks

### PR DESCRIPTION
**Pull request by amrbashir**
closes #6617

TODOs:
- [x] JS docs
 

### What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Docs
- [ ] New Binding issue #___
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

### Does this PR introduce a breaking change?

- [ ] Yes, and the changes were approved in issue #___
- [x] No

### Checklist
- [ ] When resolving issues, they are referenced in the PR's title (e.g `fix: remove a typo, closes #___, #___`)
- [x] A change file is added if any packages will require a version bump due to this PR per [the instructions in the readme](https://github.com/tauri-apps/tauri/blob/dev/.changes/readme.md).
- [ ] I have added a convincing reason for adding this feature, if necessary

### CODE REVIEW

**Comment by amrbashir** [comment]
Now that the implementation is done without depending on the `window` plugin, I still believe the `window` plugin (plus some others) should be added back to core, also notice that the `tray` plugin is dependent on the `menu` plugin so both will need to be in coupled together.

Also the `menu` plugin ideally should use the `Window` and `Position` types from the `window` plugin and the `window` plugin should depend on the `menu` plugin so that we can provide a window config option to use a specific menu when creating the window.

All this could be solved by just adding them all into one single package in core. These are essential to any tauri app and so should be provided by default in core.

File: core/tauri-macros/src/lib.rs
Comment (lucasfernog) [inline core/tauri-macros/src/lib.rs]:
```suggestion
/// Accepts a closure-like syntax to call arbitrary code on a menu item
```

File: core/tauri-macros/src/lib.rs
Comment (lucasfernog) [inline core/tauri-macros/src/lib.rs]:
```suggestion
/// after matching against `kind` and retrieving it from `resources_table` using `rid`.
```

File: core/tauri-macros/src/lib.rs
Comment (lucasfernog) [inline core/tauri-macros/src/lib.rs]:
```suggestion
/// You can optionally pass a third parameter to select which item kinds
```

File: core/tauri-macros/src/lib.rs
Comment (lucasfernog) [inline core/tauri-macros/src/lib.rs]:
```suggestion
/// to match against, by providing a `|` separated list of item kinds
```

File: core/tauri-macros/src/menu.rs
Comment (lucasfernog) [inline core/tauri-macros/src/menu.rs]:
```suggestion
  let has_negated = kinds.iter().any(|n| n.0);
```

File: core/tauri-macros/src/menu.rs
Comment (lucasfernog) [inline core/tauri-macros/src/menu.rs]:
```suggestion
    .filter_map(|nident| {
      if nident.0 {
        None
      } else {
        match nident.1 {
          i if i == "MenuItem" => Some((i, Ident::new("MenuItem", Span::call_site()))),
          i if i == "Submenu" => Some((i, Ident::new("Submenu", Span::call_site()))),
          i if i == "Predefined" => Some((i, Ident::new("PredefinedMenuItem", Span::call_site()))),
          i if i == "Check" => Some((i, Ident::new("CheckMenuItem", Span::call_site()))),
          i if i == "Icon" => Some((i, Ident::new("IconMenuItem", Span::call_site()))),
          _ => None,
        }
      }
    })
```

File: core/tauri-macros/src/menu.rs
Comment (lucasfernog) [inline core/tauri-macros/src/menu.rs]: I don't like how it assumes `kind`, `rid` and `resources_table` exists instead of taking them as arguments.. but this is internal so it's ok.

File: core/tauri/src/resources/mod.rs
Comment (lucasfernog) [inline core/tauri/src/resources/mod.rs]:
```suggestion
/// Resources can be created in commands. Resources can also be retrieved in commands by
```

File: core/tauri/src/resources/mod.rs
Comment (lucasfernog) [inline core/tauri/src/resources/mod.rs]: let's resolve this :)

File: core/tauri/src/resources/mod.rs
Comment (lucasfernog) [inline core/tauri/src/resources/mod.rs:98]:
```suggestion
  /// Returns an unique resource ID, which acts as a key for this resource.
```

File: core/tauri/src/resources/mod.rs
Comment (lucasfernog) [inline core/tauri/src/resources/mod.rs]:
```suggestion
  /// this function returns [`Error::BadResourceId`].
```

File: core/tauri/src/resources/mod.rs
Comment (lucasfernog) [inline core/tauri/src/resources/mod.rs]: should we delete this?

File: core/tauri/src/resources/plugin.rs
Comment (lucasfernog) [inline core/tauri/src/resources/plugin.rs:14]: this is problematic, an attacker can close all resources by iterating on IDs until an error is returned.

maybe we should make the ID random instead?

File: core/tauri/src/menu/plugin.rs
Comment (lucasfernog) [inline core/tauri/src/menu/plugin.rs:1]: this file is kinda big, can we move the commands to their own module?

File: core/tauri/src/tray/plugin.rs
Comment (lucasfernog) [inline core/tauri/src/tray/plugin.rs]: I think we should use a channel here instead, so we know only the tray icon caller have access to these events

also simplifies the JS APi a lot :)

File: core/tauri/src/menu/plugin.rs
Comment (lucasfernog) [inline core/tauri/src/menu/plugin.rs]: consider using a channel instead (maybe have a dedicated list of channels in the resources table)

**Review by lucasfernog — CHANGES_REQUESTED** [review]

File: core/tauri-macros/src/menu.rs
Comment (amrbashir) [reply core/tauri-macros/src/menu.rs]: I srsly don't know what I was thinking writing that code, just kill me already 😅

File: core/tauri/src/resources/mod.rs
Comment (amrbashir) [reply core/tauri/src/resources/mod.rs]: I left it there in case we ever need but I guess we can re-implement it

File: core/tauri-macros/src/menu.rs
Comment (amrbashir) [reply core/tauri-macros/src/menu.rs]: yeah I don't like it either, maybe I need to spend some time thinking about improving it without adding much noise in the call site

File: core/tauri/src/resources/plugin.rs
Comment (amrbashir) [reply core/tauri/src/resources/plugin.rs:14]: Closing resource is not that bad right?

Also for reference, deno is using the same ID strategy and though they also provide `.close()` you can't really invoke the underlying mechanism of it (compared to `invoke("plugin:resources|close"`). 

So in order to simplify the implementation and keep using incrementing ID numbers, we can hide the invoke endpoint and randomize it on each launch of the app.

File: core/tauri/src/menu/plugin.rs
Comment (amrbashir) [reply core/tauri/src/menu/plugin.rs:1]: the commands make most of the file, so moving them into another module will still be a big file

**Comment by amrbashir** [comment]
I think this PR needs to wait until the refactors PR are merged in.

File: core/tauri/src/resources/mod.rs
Comment (FabianLars) [inline core/tauri/src/resources/mod.rs]: idk why the github notification showed me a single change (`a` to `an`) and literally nothing else, but that's the only reason i noticed it i guess. `a unique` was correct because it starts with a `yu` sound.
And people say german is complicated... 😡 
```suggestion
  /// Returns a unique resource ID, which acts as a key for this resource.
```

**Comment by lucasfernog** [comment]
We're missing a lot of documentation here, but I wanna merge it anyway because we need to do a complete pass on API examples.

I like how this can be dynamic with the MenuItem classes (so you can update, share items), or static with raw objects after https://github.com/tauri-apps/tauri/pull/7709/commits/d0eeaba4198c3b04b5f252bd87f5ff87b3359fe2

**Review by lucasfernog — APPROVED** [review]

**Pull request merged**
